lib/submit.js

```diff
@@ -0,0 +1,122 @@
+import { NODE_PROXY_URI } from './constants'
+
+/**
+ * Submit hash(es) to one or more Nodes, returning an Array of proof handle objects, one for each submitted hash and Node combination.
+ * @param {Array<String>} hashes - An Array of String Hashes in Hexadecimal form.
+ * @param {Array<String>} uris - An Array of String URI's. Each hash will be submitted to each Node URI provided. If none provided three will be chosen at random using service discovery.
+ * @return {Array<{uri: String, hash: String, hashIdNode: String, groupId: String}>} An Array of Objects, each a handle that contains all info needed to retrieve a proof.
+ */
+async function submitHashes(hashes, uris) {
+  uris = uris || []
+  let nodes
+
+  // Validate args before doing anything else
+  validateHashesArg(hashes, h => isHex(h))
+  validateUrisArg(uris)
+
+  if (isEmpty(uris)) {
+    // get a list of nodes via service discovery
+    nodes = await getNodes(3)
+  } else {
+    // eliminate duplicate URIs
+    uris = uniq(uris)
+
+    // non-empty, check that *all* are valid or throw
+    let badURIs = reject(uris, function(h) {
+      return isValidNodeURI(h)
+    })
+    if (!isEmpty(badURIs)) throw new Error(`uris arg contains invalid URIs : ${badURIs.join(', ')}`)
+    // all provided URIs were valid
+    nodes = uris
+  }
+
+  try {
+    // Setup an options Object for each Node we'll submit hashes to.
+    // Each Node will then be sent the full Array of hashes.
+    let nodesWithPostOpts = map(nodes, node => {
+      let uri = isSecureOrigin() ? NODE_PROXY_URI : node
+      let headers = Object.assign(
+        {
+          'Content-Type': 'application/json',
+          Accept: 'application/json'
+        },
+        isSecureOrigin()
+          ? {
+              'X-Node-Uri': node
+            }
+          : {}
+      )
+
+      let postOptions = {
+        method: 'POST',
+        uri: uri + '/hashes',
+        body: {
+          hashes: hashes
+        },
+        headers,
+        timeout: 10000
+      }
+      return postOptions
+    })
+
+    // All requests succeed in parallel or all fail.
+    const parsedBody = await promiseMap(nodesWithPostOpts, fetch, {
+      concurrency: 25
+    })
+    // Nodes cannot be guaranteed to know what IP address they are reachable
+    // at, so we need to amend each result with the Node URI it was submitted
+    // to so that proofs may later be retrieved from the appropriate Node(s).
+    // This mapping relies on that fact that promiseMap returns results in the
+    // same order that options were passed to it so the results can be mapped to
+    // the Nodes submitted to.
+    forEach(nodes, (uri, index) => {
+      parsedBody[index].meta.submitted_to = uri
+    })
+
+    // Map the API response to a form easily consumable by getProofs
+    let proofHandles = mapSubmitHashesRespToProofHandles(parsedBody)
+
+    return proofHandles
+  } catch (err) {
+    console.error(err.message)
```

Comment:
Is there any reason to keep these console.error's? If we're just throwing on an error anyway, it feels like it defeats the purpose of the `try/catch` and we should just let any errors just bubble up normally, leaving it up to the consumer/client to `try/catch` and deal with any errors.

Reply:
I'm in the process of refactoring this from `resolve/reject` to `async/await` in other files now too and felt worth asking.

Reply:
In this particular case, I see no reason why we must catch the errors at the point we are currently doing. If we change this, we'll just want to be sure that we inspect and possibly adjust our error handling on anything we have that consumes this, such as the `cli`.

Reply:
Yep! Agreed. Ok, so I'll refactor these out and then start testing client with cli once this refactor is done.